### ISSUE TITLE
B007: False positive if loop control variable is used after the loop

### ISSUE THREAD

**Issue by jdufresne** [PyCQA/flake8-bugbear#68]
Example code:

```py
from mymodule import myiter


def main():
    for result in myiter():
        pass
    print(result)
```

Ouptut:

```
foo.py:6:9: B007 Loop control variable 'result' not used within the loop body. If this is intended, start the name with an underscore.
```

This code is intentionally exhausting an iterator without generating a temporary data structure. As the loop control variable is used, I believe this should not emit a warning.

**Comment by cooperlees** [comment]
I find this logic easy to miss and have personally seen bad bugs and outages caused because if it. 

I believe being more explicit in your code that you want to use the variable outside the loop makes it much harder to miss. For example:
```
from mymodule import myiter


def main():
    last_object = None
    for result in myiter():
        last_object = result
    print(last_object)
```

On these grounds, I vote for it to stay. As always, open to further discussion tho.

**Comment by jdufresne** [comment]
I dunno, IMO, that proposed form adds multiple unnecessary assignments. I recognize this may simply be my opinion and it is subjective. Thanks for considering.

> Another benefit here is that you will also just get "None" printed and not throw an exception if result is never set ...

This is trivial to solve with original form too:

```py
    result = None
    for result in myiter():
        pass
    print(result)
```

**Comment by cooperlees** [comment]
I understand, but I feel `bugbear` is designed to be opinionated, and I like it's current opinion.

If you do the latter suggestion, I feel the cost of the extra assignments are so negligible that you can either:
- Do the extra assignment with a descriptive variable name
- Add the # noqa: B007 as it shows that you're doing a *risky* thing and stands out, AND, you don't get the assignment cost you're worried about.

Please reopen if you strongly disagree, but I feel this is all acceptable. Once again, all ears.

**Comment by kerrick-lyft** [comment]
Here's another case that I think is valid and idiomatic, but `flake8-bugbear` reports a B007 warning that `idx` is unused:

```
# Print the slice of nums starting from the first multiple of 7 until the end, or an empty list if there are no multiples of 7.
nums = list(range(10, 20))

idx = -1
for idx, num in enumerate(nums):
    if num % 7 == 0:
        break

print(nums[idx:])
```

**Comment by cooperlees** [comment]
For that specific case, where the variable has a value before the loop, I'd accept a fix. Don't know how you'd implement it sanely tho (I'm no AST god tho).

**Comment by Nodd** [comment]
I'm bitten by the same problem. Could the check be changed from "Loop control variable not used within loop body" to "Loop control variable not used", as this seems to be the real problem to be checked here ?

Here is my relevant code, where I'm explicitly looking for the index:
```python
        for i, deriv in enumerate(data_db_smooth_diff):
            if deriv < 10e-8:
                break
        if i:
            data_db_smooth[:i] = data_db_smooth[i]
```
I can use the workaround shown above, but the resulting code is honestly not that great.

**Comment by jaraco** [comment]
> I find this logic easy to miss and have personally seen bad bugs and outages caused because if it.

Can you give an example of how a bug came out of using the loop variable outside the loop that was caused by using the loop variable solely outside the loop?

**Comment by cooperlees** [comment]
There are two main ways here I've see issues for using loop control variables undesired ways:

- Used out of sight (outside the loop) and the loop logic was changed / what it iterates causing undesired behavior
- accidentally overwriting a variable defined outside the loop (kinda of related here)

It's caused pain in larger python files where the variable is used is very far away (line wise) from the loop where it was either created or accidentally overwritten.

```python
def foo():
    bar = False
    ... # lots of code
    for i, bar in enumerate(some_list):
         do_stuff(bar)
    ... # lots of code
    if bar:
        delete_the_world()
```

Sure, good unittesting should catch this, but i think bugbear helps someone know they could be doing something evil here. if not, add the ignore and the developer is more explicit (was my thought why to add this to bugbear).

Can you explain why you wanted me to give an example? How has my example helped here?
